src/main/java/org/swisspush/redisques/handler/GetQueuesItemsCountHandler.java

```diff
@@ -1,103 +1,148 @@
+import static org.swisspush.redisques.util.RedisquesAPI.STATUS;
 
 public class GetQueuesItemsCountHandler implements Handler<AsyncResult<Response>> {
 
     private final Logger log = LoggerFactory.getLogger(GetQueuesItemsCountHandler.class);
 
+    private final Vertx vertx;
     private final Message<JsonObject> event;
     private final Optional<Pattern> filterPattern;
     private final String queuesPrefix;
     private final RedisProvider redisProvider;
+    private final UpperBoundParallel upperBoundParallel;
+    private final Semaphore redisRequestQuota;
 
     public GetQueuesItemsCountHandler(
+            Vertx vertx,
             Message<JsonObject> event,
             Optional<Pattern> filterPattern,
             String queuesPrefix,
-            RedisProvider redisProvider
+            RedisProvider redisProvider,
+            Semaphore redisRequestQuota
     ) {
+        this.vertx = vertx;
         this.event = event;
         this.filterPattern = filterPattern;
         this.queuesPrefix = queuesPrefix;
         this.redisProvider = redisProvider;
+        this.upperBoundParallel = new UpperBoundParallel(vertx);
+        this.redisRequestQuota = redisRequestQuota;
     }
 
     @Override
     public void handle(AsyncResult<Response> handleQueues) {
-        if (handleQueues.succeeded()) {
-            List<String> queues = HandlerUtil.filterByPattern(handleQueues.result(),
-                filterPattern);
-            if (queues.isEmpty()) {
-                log.debug("Queue count evaluation with empty queues");
-                event.reply(new JsonObject().put(STATUS, OK).put(QUEUES, new JsonArray()));
-                return;
-            }
-
-            redisProvider.connection().onSuccess(conn -> {
-                List<Future> responses = queues.stream()
-                        .map(queue -> conn.send(Request.cmd(Command.LLEN, queuesPrefix + queue)))
-                        .collect(Collectors.toList());
-                CompositeFuture.all(responses).onFailure(ex -> {
+        if (!handleQueues.succeeded()) {
+            log.warn("Concealed error", new Exception(handleQueues.cause()));
+            event.reply(new JsonObject().put(STATUS, ERROR));
+            return;
+        }
+        var ctx = new Object() {
+            Redis redis;
+            Iterator<String> iter;
+            List<String> queues = HandlerUtil.filterByPattern(handleQueues.result(), filterPattern);
+            int iNumberResult;
+            int[] queueLengths; /*TODO consider using primitive type*/
+        };
+        if (ctx.queues.isEmpty()) {
+            log.debug("Queue count evaluation with empty queues");
+            event.reply(new JsonObject().put(STATUS, OK).put(QUEUES, new JsonArray()));
+            return;
+        }
+        if (redisRequestQuota.availablePermits() <= 0) {
+            event.reply(new NoStackReplyException(ReplyFailure.RECIPIENT_FAILURE, 507, "Server too busy. Rejecting this GetQueuesItemsCount request"));
```

Comment:
We could also return a https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

Reply:
The correct error status is 429, because the client should handle it and throttle on its side. A 5xx indicates that the client can do nothing clever to overcome the problem,

Reply:
Applied :)